### ISSUE TITLE
PreventGradients in SoftmaxCrossEntropyWithLogit ops

### ISSUE THREAD

**Issue by renmengye** [tensorflow/tensorflow#9368]
SparseSoftmaxCrossEntropyWithLogits cannot take second order gradients (It's not a beautiful hack, and I am not sure how much computational speed up it will bring). 
Adding PreventGradient node in the gradient graph seems to contaminate the computation graph structure, e.g. I am doing custom forward-mode automatic differentiation.
Reference: https://github.com/tensorflow/tensorflow/blob/r1.0/tensorflow/python/ops/nn_grad.py#L334

**Comment by drpngx** [comment]
@renmengye I am not sure what you are asking? Is the node preventing from attempting autodiff?

CC: @vrv

**Comment by renmengye** [comment]
For example, I cannot take tf.gradients() on a gradient node to compute second order derivatives. 
y = sparse_softmax_cross_entropy_with_logits(logits=z, labels=y_)
grad_z = tf.gradients(y, z)
grad_grad_z = tf.gradients(y, grad_z) # This returns [None], thus no second order derivative can be taken.

**Comment by michaelisard** [comment]
@girving any ideas?

**Comment by girving** [comment]
The solution is to port the trick from the nonsparse case (https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/nn_grad.py#L393).

**Comment by MichaelKonobeev** [comment]
I'm trying to port the nonsparse solution to this case. Seems easy enough for my first contribution :)

**Comment by ebrevdo** [comment]
This had to be rolled back, unfortunately, because in eager mode the gradient computation accidentally inserts hessian gradients as it can't tell that the input op is a Zeros or ZerosLike.  We'll have to fix this first.

**Comment by ebrevdo** [comment]
We have an internal solution for better handling of second derivatives with `tf.custom_gradient`; will try modifying the non-sparse xent gradient to use that and report back here.

**Comment by dynamicwebpaige** [comment]
Thanks, @ebrevdo! Any updates on second derivatives with `tf.custom_gradient`?

**Comment by tilakrayal** [comment]
@renmengye,
I can see there are multiple PR's raised and merged for this feature. With the PR's #22231 and #31553, the changes were happened and again those were rolled back because in eager mode the gradient computation accidentally inserts hessian gradients as it can't tell that the input op is a Zeros or ZerosLike. 

With the PR #31700, the Hessian was implemented for the sparse softmax cross entropy and also find the changes in the `nn_grad.py` file

https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/nn_grad.py#L539

Thank you!
